enhancements/update/change-management-and-maintenance-schedules.md

```diff
@@ -0,0 +1,1273 @@
+    the control-plane and for worker machine config. They can also see the date and time that the next material change will be permitted.
+    The MCP will not show a pending change at this time, but will show the next time at which material changes will be permitted.
+13. The next Saturday is _not_ the first Saturday of the month. The CVO detects that material changes are permitted at 00:00 UTC and
+    begins to apply manifests. This effectively initiates the control-plane update process, which is considered a single
+    material change to the cluster.
+14. The control-plane update succeeds. The CVO, having reconciled its state, unsets metrics suggesting changes are pending.
+15. As part of updating cluster manifests, MachineConfigs have been modified. The MachineConfigOperator (MCO) re-renders a
+    configuration for worker-nodes. However, because the MCP maintenance schedule precludes initiating material changes,
+    it will not begin to update Machines with that desired configuration.
+16. The MCO will set a metric indicating that desired changes are pending. 
+17. `oc get -o=yaml/describe` will both provide status information indicating that changes are pending for the MCP and
+    the time at which the next material changes can be initiated according to the maintenance schedule.
+18. On the first Saturday of the next month, 00:00 UTC, the MCO determines that material changes are permitted.
+    Based on limits like maxUnavailable, the MCO begins to annotate nodes with the desiredConfiguration. The 
+    MachineConfigDaemon takes over from there, draining, and rebooting nodes into the updated release.
+19. There are a large number of nodes in the cluster and this process continues for more than 24 hours. On Saturday
+    23:59, the MCO applies a round of desired configurations annotations to Nodes. At 00:00 on Sunday, it detects
+    that material changes can no longer be initiated, and pauses its activity. Node updates that have already
+    been initiated continue beyond the maintenance schedule window.
+20. Since not all nodes have been updated, the MCO continues to expose a metric informing the system of
+    pending changes.
+21. In the subsequent days, the cluster is scaled up to handle additional workload. The new nodes receive
+    the most recent, desired configuration.    
+22. On the first Saturday of the next month, the MCO resumes its work. In order to ensure that forward progress is
+    made for all nodes, the MCO will update nodes that have the oldest current configuration first. This ensures
+    that even if the desired configuration has changed multiple times while maintenance was not permitted,
+    no nodes are starved of updates. Consider the alternative where (a) worker-node updates required > 24h,
+    (b) updates to nodes are performed alphabetically, and (c) MachineConfigs are frequently being changed

```

Comment:
If the maint. windows were small enough, and changes often enough, I guess it's conceivable you end up with say, 10 worker nodes on 10 different versions, and you never have a completely consistent pool.

I guess version skew guards will prevent any of these from actually failing, but we should probably report status such that a user can understand that thats going on.

Reply:
That is a potential outcome. As you suggest, I don't think it is dangerous as long as we have skew control.

Reply:
Assuming a new version comes along later, v3, will the MCO recognise that 10 are on v1, 10 are on v2, and update those on v1 first, or will it be random? Then you might end up with a mix of v1, v2 and v3?

What is our reporting like for understanding which version a Node is on? I could imagine a situation where we get a bug and the customers says yes, cluster is on version X.Y.Z, and we can see a bug there that was fixed in say X.Y.Z-1, but their nodes are mixed across several versions prior to that, how do we make it clear that the node that broke in this instance is on an unpatched release and therefore the customer needs to upgrade to fix the bug?

And that said, what if there's a change that affects the rendered config, that also relies on controllers being updated, or vice versa, how do we handle skew there 🤔